docs/design/gluster-block-provisioning.md

```diff
@@ -0,0 +1,350 @@
+# Gluster-Block Provisioning End-To-End Design
+
+## Motivation
+
+While GlusterFS volumes can be used for RWO as well as RWX (and ROX) volumes,
+for some workloads, specifically metadata-heavy, the RWO support is not
+satisfactory with respect to performance. Furthermore, the scalability
+in number of volumes is currently not very good.
+
```

Comment:
Thanks for your review, @wattsteve!

I am happy to clarify the text a bit, but here are a few thoughts. Maybe it's me who has not really understood the problem we want to solve:

* RWO is defined as "the volume can be mounted as read-write by a single node": see https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes. (This sounds like a very very strange mode, because you can have as many writers as you want as long as they are on the same node. But even with that strange definition, I heard that this separation is not enforced by Kubernetes in any way.)
* Reading the definitions, a storage volume that can even be used for RWX purposes, can as well be exported as a RWO volume. Conceptually, there is nothing wrong about this. Hence we do provide perfectly valid RWO volumes with GlusterFS today.
* Also many users are using GlusfterFS backed RWO volumes today. So saying, we don't support RWO seems wrong to me.
* But for all I know, it is more the **expectations** that the users of RWO volumes have, specifically with respect to **performance**, that are not met by our GlusterFS-backed RWO volumes. This is because that capability of supporting multiple concurrent writers comes at a big cost for metadata-performance. So some users are happy with the performance, but others with more RWO-specific workloads are suffering from the performance. _This was the first and foremost reason for us to propose this design._
* I don't understand the comment about a RWX volume not providing the single-writer or locking semantics that you mention:
    * The file system of a RWO volume (i.e. with a single writer) would not need to care about locking as much as a RWX/shared volume where locking is crucially important. So what locking semantics does a RWO volume provide?
    * The file system itself usually can't do anything much about single-writer, right? This needs to be enforced by the kubernetes level if we want to do it, or what am I missing? (E.g. for the iSCSI mount plugin @humblec recently added the strict single-writer protection so that the lack of kube-wide enforcement mechanisms does not destroy our data...)